### PR TITLE
DRAFT: counting with longs

### PR DESCRIPTION
**Pull request by aaime**
Draft PR to verify how much things change by counting using longs instead of ints.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [ ] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [ ] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [ ] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: modules/library/main/src/main/java/org/geotools/data/FeatureSource.java [hunk 1]
```diff
@@ -206,8 +206,36 @@
      *         or {@code -1} if this cannot be calculated.
      *
      * @throws IOException if there are errors getting the count
+     * @deprecated Please use {@link #count(Query)} instead
      */
-    int getCount(Query query) throws IOException;
+    @Deprecated
+    default int getCount(Query query) throws IOException {
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/data/FeatureSource.java:212]: Document use of Integer.MAX_VALUE:
```
If this source contains more than Integer.MAX_VALUE elements, returns Integer.MAX_VALUE.
```

File: modules/library/main/src/main/java/org/geotools/data/FeatureSource.java [hunk 2]
```diff
@@ -206,8 +206,36 @@
...
+     * start index set on the {@code Query}.
+     * <p>
+     * It is possible that this method will return {@code -1} if the calculation
+     * of number of features is judged to be too costly by the implementing class.
+     * In this case, you might call <code>getFeatures(query).size()</code>
+     * instead.
+     * <p>
+     * Example use:<pre><code> int count = featureSource.getCount();
+     * if( count == -1 ){
+     *    count = featureSource.getFeatures( "typeName", count ).size();
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/data/FeatureSource.java:228]: Is a cast or `Math.toIntExact` needed (assuming `count` is a long here).

```
      *  if( count == -1 ){
      *    count = Math.toIntExact( featureSource.getFeatures( "typeName", count ).size() );
```

**Review by jodygarnett — APPROVED** [review]

File: modules/library/main/src/main/java/org/geotools/data/FeatureSource.java (hunk 1 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/data/FeatureSource.java:212]: I think it should throw an exception instead?

File: modules/library/main/src/main/java/org/geotools/data/FeatureSource.java (hunk 2 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/data/FeatureSource.java:228]: Yes, this is the result of a "inline method" refactor, I did not check call point by call point, the PR is most to try and get some grip on the discussion on gt-devel. The way it's going, it's not very much compatible with my timeframes, the DGGS activity is just a prototype at the moment and is going to end soon, I was hoping to get some useful API improvement out of it, on the side.

File: modules/library/main/src/main/java/org/geotools/data/FeatureSource.java (hunk 1 above)
Comment (mprins) [reply modules/library/main/src/main/java/org/geotools/data/FeatureSource.java:212]: If throwing it should really be [`java.lang.ArithmeticException`](https://docs.oracle.com/javase/8/docs/api/java/lang/ArithmeticException.html) to stay in line with what happens in eg. `java.lang.Math` when the result overflows an int, but method only declares `IOException` which seems counter-intuitive to throw. `ArithmeticException` being unchecked doesn't really need to be declared of course, but the unwitting programmer could be in for a surprise in that case.
That being said, I would prefer if it throws anything to having it return a magic value (even if it is `IOException` wrapping `ArithmeticException`)

**Pull request closed without merging**